### ISSUE TITLE
Refactoring the class structure for postprocessing

### ISSUE THREAD

**Issue by wiederm** [choderalab/modelforge#180]
As @chrisiacovella mentioned [here](https://github.com/choderalab/modelforge/issues/177), the postprocessing module should be restructured. I am opening a new issue because we should include a few additional points not part of the original issue and increase its scope.

In the following, I will go ahead and outline the parts that we want to refactor and suggest what a successful PR should accomplish. All of the following is part of the property prediction postprocessing.

I think that it is necessary to provide as an input to the module (and to the postprocessing module) the property names of the properties that we want to predict with a model, the output dimension of a property (indicating tensor or vector property), and the property name this corresponds to in the dataset. This information must be passed to the model and will be used to customize the output heads, register the loss functions, and retrieve the loss scaling factor, it will be used to log the loss (the name of the loss) and define which postprocessing and reduction operations should be performed.

Customizable output heads
--------------------------------------
Each each model is trained on a scalar (energy) and a vector (force) per default. This is done using a single output head (predicting molecular energies and calculating the force). We should add the option to add additional output heads that can be trained on scalar or vector properties (note that this is not possible for every model architecture, and some architectures (like PAINN, Tensornet, and PhysNet) already output additional properties.

I think there are two PRs that we should open to accomplish two goals:

**High priority**: Provide the option to define the output properties (number of properties and scalar or vector) at runtime via the toml files. Only the infrastructure to determine these is required, no model modifications needed at that point.

**Low priority**: Modify the models to add the flexibility. This is trivial for some models, e.g., for SchNet, defining the output size as (batch_size, 2) allows it to train on 2 scalar properties. 

Loss function modularity
-----------------------------------------

Register the output properties in the `Loss` module. We can use the same loss function (squared L2 norm) for vector and scalar properties (as we do now). The `Loss` module calculates the loss as the sum of the individual losses using a weight factor for individual properties. Ideally, we want to divide extensive properties (which depend on the number of atoms) by the number of atoms to avoid weighting energies by the number of atoms of the molecule.

A successful PR should refactor the current Loss module and register a loss for each predicted property.

Logging for properties
-----------------------------------------

Related to the loss module:
A logging module should register the different properties (or the loss itself?) for which loss functions are registered and should log these. The metrics we base decision on are validation set energy RMSE, these are being registered separately, here we are only concerned about logging the loss metrics.

Postprocessing of atomic properties
----------------------------------------------------------------------------------

For each requested property we can define postprocessing and reduction operations in the toml file. I think that the best way to do this is to define for each property the operation that should be performed (e.g., `scale`, `calculate_self_energy`, `from_atom_to_molecule_reduction`) and provide potential keywords This is taking up the discussion we have started in [issue 177](https://github.com/choderalab/modelforge/issues/177). I think I am backtracking some of what I have written in that issue. Thinking through the full pipeline I think we should not tie operations to properties (i.e., we shouldn't call it `scale_energy` but just `scale_property`, and it seems better to provide mean and stddev. If we think that this is not robust (and I agree with that), because we are providing only one kind of mean and stddev in the module, I think it is much better to provide these values in the toml file.

A successful PR should outline the options for the toml file that can be passed to the `PostProcessing` module defining the operations that should be performed at runtime.

**Comment by wiederm** [comment]
Closing this because the discussion in PR #177 has picked up all the raised points.
